File: api/bun-notes.py
"""
/api/bun-notes — the note store behind /football/bakers-buns.

GET    : every note, oldest first. `?team=CHI` narrows to one team, `?team=` to
         the league-wide ones. Public, because the notes are published reading
         for everyone; the page renders them for signed-out visitors too.
PUT    : create or update. Body is one note object or {"notes": [...]} so the
         modal's multi-bullet save is a single request. Any active account may
         write; an existing note needs its author or an admin.
DELETE : ?id=n_x, one note. Its author or an admin.

Storage: a Supabase table `bun_notes(id text pk, team, week, data jsonb, ...)`.
See scripts/sql/bun_notes.sql for the schema and scripts/seed_bun_notes.py for
the one-time import of the notes that were seeded from the Google Doc.
"""
from http.server import BaseHTTPRequestHandler
from datetime import datetime, timezone
import json
import os
import re
import urllib.parse
import urllib.request

SUPABASE_URL = os.environ.get("SUPABASE_URL", "")
SUPABASE_KEY = os.environ.get("SUPABASE_KEY", "")

# The one combined picker: a general note, the preseason, the 18 regular-season
# weeks, then the four playoff rounds. Anything else is rejected rather than
# stored, so the page never has to render a week label it has no name for.
WEEKS = {"all", "pre", "wc", "div", "conf", "sb"} | {str(n) for n in range(1, 19)}
KINDS = {"note", "schedule"}
MAX_TEXT = 2000
MAX_BATCH = 50

# Ids are minted by the client ('n_...') or by the seeder ('seed_CHI_note_0'),
# so they are only ever this alphabet. Pinning that down is also what makes them
# safe to interpolate into the `id=in.(...)` filter the ownership pre-read uses.
NOTE_ID = re.compile(r"^[A-Za-z0-9_-]{1,64}$")


def supabase_request(path, method="GET", body=None, headers=None):
    url = f"{SUPABASE_URL}/rest/v1/{path}"
    data = json.dumps(body).encode() if body is not None else None
    req = urllib.request.Request(url, data=data, method=method)
    req.add_header("apikey", SUPABASE_KEY)
    req.add_header("Authorization", f"Bearer {SUPABASE_KEY}")
    req.add_header("Content-Type", "application/json")
    for k, v in (headers or {}).items():
        req.add_header(k, v)
    with urllib.request.urlopen(req) as resp:
        raw = resp.read()
        return json.loads(raw) if raw else None


def fetch_user(user_id):
    rows = supabase_request(
        f"users?id=eq.{urllib.parse.quote(user_id)}&select=role,status")
    return rows[0] if rows else None


def now_iso():
    return datetime.now(timezone.utc).isoformat()


# ---- who may write what -----------------------------------------------------
# The notes are a shared, published store: anyone signed in may file one, and
# owns what they filed. An admin owns everything. These three live at module
# level rather than on the handler because server.py loads this file to mirror
# the rule locally — see bun_notes_api() there.


def resolve_actor(user_id):
    """Returns (user_id, is_admin, error), error being None or (status, body).

    Any active account may write notes; the admin flag decides whose notes they
    may touch besides their own.
    """
    if not user_id:
        return None, False, (401, {"error": "Not authenticated"})
    user = fetch_user(user_id)
    if not user or user.get("status") is not True:
        return None, False, (403, {"error": "Account is inactive"})
    return user_id, user.get("role") == "admin", None


def fetch_authors(ids):
    """Map of note id -> authorId, for the ids that already exist.

    One query for the whole batch. A missing key means the note is new; a None
    value means an unowned note — the seeded import filed those with no author
    (see scripts/seed_bun_notes.py), so only an admin can touch them.
    """
    ids = [i for i in ids if i]
    if not ids:
        return {}
    quoted = ",".join('"' + i + '"' for i in ids)
    rows = supabase_request(
        "bun_notes?id=in.(" + urllib.parse.quote(quoted) + ")&select=id,data")
    return {r["id"]: (r.get("data") or {}).get("authorId") for r in (rows or [])}


def may_write(authors, note_id, user_id, is_admin):
    """None when the write is allowed, else the message to 403 with.

    A note whose id is not in `authors` does not exist yet, and creating one is
    always allowed.
    """
    if is_admin or note_id not in authors:
        return None
    if authors[note_id] and authors[note_id] == user_id:
        return None
    return "You can only edit your own notes"


def clean_source(src):
    """Keep only the two fields the card renders, and only for a real x.com post.

    The client parses the handle out of the URL and we re-parse it here rather
    than trusting it, so a stored source can never render a label that points
    somewhere other than the link.
    """
    if not isinstance(src, dict):
        return None
    url = (src.get("url") or "").strip()
    if not url:
        return None
    parsed = urllib.parse.urlparse(url)
    if parsed.scheme not in ("http", "https"):
        return None
    host = (parsed.netloc or "").lower()
    if host.startswith("www."):
        host = host[4:]
    if host not in ("x.com", "twitter.com"):
        return None
    parts = [p for p in parsed.path.split("/") if p]
    # /<handle>/status/<id>
    if len(parts) < 3 or parts[1] not in ("status", "statuses") or not parts[2].isdigit():
        return None
    return {"url": url, "handle": parts[0]}


def note_id_of(note):
    """Pull the id off a note from the wire. Returns (id, error_message).

    Split out of normalize() because the ids have to be read and checked before
    anything else: they are what the ownership pre-read asks about, and its
    answer is what normalize() then stamps as the author.
    """
    if not isinstance(note, dict):
        return None, "note must be an object"
    note_id = (note.get("id") or "").strip()
    if not note_id:
        return None, "note id is required"
    if not NOTE_ID.match(note_id):
        return None, f"invalid note id {note_id!r}"
    return note_id, None


def normalize(note, author_id):
    """Validate one note off the wire. Returns (row, error_message).

    `author_id` is the note's owner as resolved by the caller: the requester for
    a new note, the original author for an edit.
    """
    note_id, error = note_id_of(note)
    if error:
        return None, error

    text = (note.get("text") or "").strip()
    if not text:
        return None, "note text is required"
    if len(text) > MAX_TEXT:
        return None, f"note text is longer than {MAX_TEXT} characters"

    week = str(note.get("week") or "all")
    if week not in WEEKS:
        return None, f"unknown week {week!r}"

    kind = note.get("kind") or "note"
    if kind not in KINDS:
        return None, f"unknown kind {kind!r}"

    team = (note.get("team") or "").strip().upper()

    data = {
        "id": note_id,
        "team": team,
        "week": week,
        "kind": kind,
        "text": text,
        # The author and the timestamps are stamped here, not accepted from the
        # client. Both the author and createdAt are carried forward on an edit
        # rather than restamped: an admin correcting someone's note must not
        # take it off them, and an edit keeps its original date.
        "authorId": author_id,
        "createdAt": note.get("createdAt") or now_iso(),
        "updatedAt": now_iso(),
    }
    source = clean_source(note.get("source"))
    if source:
        data["source"] = source
    if isinstance(note.get("order"), int):
        data["order"] = note["order"]
    if note.get("seeded"):
        data["seeded"] = True

    return {
        "id": note_id,
        "team": team,
        "week": week,
        "data": data,
        "updated_at": "now()",
    }, None


class handler(BaseHTTPRequestHandler):
    def _actor(self):
        return resolve_actor(self.headers.get("X-User-Id"))

    def do_GET(self):
        params = urllib.parse.parse_qs(
            urllib.parse.urlparse(self.path).query, keep_blank_values=True)
        query = "bun_notes?select=data&order=created_at"
        if "team" in params:
            team = (params["team"][0] or "").upper()
            query += "&team=eq." + urllib.parse.quote(team)
        try:
            rows = supabase_request(query)
            self._json(200, [r["data"] for r in (rows or [])])
        except Exception as e:
            self._json(500, {"error": str(e)})

    def do_PUT(self):
        user_id, is_admin, err = self._actor()
        if err:
            self._json(*err)
            return

        try:
            length = int(self.headers.get("Content-Length", 0))
            body = json.loads(self.rfile.read(length)) if length else {}
        except Exception:
            self._json(400, {"error": "Body must be JSON"})
            return

        notes = body.get("notes") if isinstance(body, dict) and "notes" in body else [body]
        if not isinstance(notes, list) or not notes:
            self._json(400, {"error": "No notes in the request"})
            return
        if len(notes) > MAX_BATCH:
            self._json(400, {"error": f"At most {MAX_BATCH} notes per request"})
            return

        ids = []
        for note in notes:
            note_id, error = note_id_of(note)
            if error:
                self._json(400, {"error": error})
                return
            ids.append(note_id)

        try:
            authors = fetch_authors(ids)
        except Exception as e:
            self._json(500, {"error": str(e)})
            return

        rows = []
        for note, note_id in zip(notes, ids):
            refused = may_write(authors, note_id, user_id, is_admin)
            if refused:
                self._json(403, {"error": refused})
                return
            # An existing note keeps the author it was filed under; a new one is
            # owned by whoever is writing it.
            row, error = normalize(note, authors.get(note_id, user_id))
            if error:
                self._json(400, {"error": error})
                return
            rows.append(row)

        try:
            # Upsert on the primary key. created_at is left out of the payload
            # so an edit keeps the row's original insert time.
            supabase_request(
                "bun_notes?on_conflict=id",
                method="POST",
                body=rows,
                headers={"Prefer": "resolution=merge-duplicates,return=minimal"},
            )
            self._json(200, {"ok": True, "notes": [r["data"] for r in rows]})
        except Exception as e:
            self._json(500, {"error": str(e)})

    def do_DELETE(self):
        user_id, is_admin, err = self._actor()
        if err:
            self._json(*err)
            return
        params = urllib.parse.parse_qs(urllib.parse.urlparse(self.path).query)
        note_id = (params.get("id", [None])[0] or "").strip()
        if not note_id:
            self._json(400, {"error": "id parameter is required"})
            return
        if not NOTE_ID.match(note_id):
            self._json(400, {"error": f"invalid note id {note_id!r}"})
            return
        try:
            # An id that matches nothing stays a 200 no-op, as it always was.
            refused = may_write(fetch_authors([note_id]), note_id, user_id, is_admin)
        except Exception as e:
            self._json(500, {"error": str(e)})
            return
        if refused:
            self._json(403, {"error": refused})
            return
        try:
            supabase_request(
                "bun_notes?id=eq." + urllib.parse.quote(note_id),
                method="DELETE",
                headers={"Prefer": "return=minimal"},
            )
            self._json(200, {"ok": True})
        except Exception as e:
            self._json(500, {"error": str(e)})

    def _json(self, status, data):
        self.send_response(status)
        self.send_header("Content-Type", "application/json")
        self.send_header("Access-Control-Allow-Origin", "*")
        self.send_header("Access-Control-Allow-Headers", "Content-Type, X-User-Id")
        self.end_headers()
        self.wfile.write(json.dumps(data).encode())

    def do_OPTIONS(self):
        self.send_response(204)
        self.send_header("Access-Control-Allow-Origin", "*")
        self.send_header("Access-Control-Allow-Methods", "GET, PUT, DELETE, OPTIONS")
        self.send_header("Access-Control-Allow-Headers", "Content-Type, X-User-Id")
        self.end_headers()

File: scripts/primary/bun-notes.js
/* Baker's Buns notes — the published note store behind /football/bakers-buns,
 * and the admin modal that writes to it.
 *
 * The notes started as a one-off import of the "NFL 2026-27" Google Doc into
 * data/nfl_projections_2026.json. They now live in Supabase (see
 * scripts/sql/bun_notes.sql and api/bun-notes.py) so they can be added to and
 * corrected from the page itself; scripts/seed_bun_notes.py moved the doc's
 * notes across, and the JSON copy stays only as the fallback this file's
 * consumer renders when the API cannot be reached.
 *
 * Reading is public. Writing is admin-only — enforced on the server, with the
 * UI here merely declining to show controls that would 403 anyway.
 *
 * Depends on: scripts/base/auth.js (getUser / isAdmin) and
 * scripts/components/nfl-pickers.js (the team grid and week menu). Styles for
 * the modal and the bullets live in styles/primary/bakers-buns.css.
 */
(function (global) {
  'use strict';

  var API = '/api/bun-notes';

  function esc(s) {
    return String(s == null ? '' : s).replace(/[&<>"']/g, function (c) {
      return { '&': '&amp;', '<': '&lt;', '>': '&gt;', '"': '&quot;', "'": '&#39;' }[c];
    });
  }

  function admin() {
    return typeof isAdmin === 'function' && isAdmin();
  }

  /* ---------- the season, as one list ----------
   * A note is filed against a week, and "the whole season" is one of the
   * answers — the default one, in fact, since most of what gets written about a
   * team isn't about a particular Sunday. The preseason and the four playoff
   * rounds sit in the same list as the eighteen numbered weeks rather than
   * behind a separate phase control: the writer picks when it happened once,
   * and there is only ever one thing to pick. `short` is what the bullet's chip
   * shows, which is why it has to stay narrow. */
  var WEEKS = [{ value: 'all', label: 'General — no week', short: '' },
               { value: 'pre', label: 'Preseason', short: 'PRE', groupStart: true }];

  for (var w = 1; w <= 18; w++) {
    WEEKS.push({ value: String(w), label: 'Week ' + w, short: 'W' + w, groupStart: w === 1 });
  }

  [['wc', 'Wild Card', 'WC'], ['div', 'Divisional', 'DIV'],
   ['conf', 'Conference', 'CONF'], ['sb', 'Super Bowl', 'SB']]
    .forEach(function (r, i) {
      WEEKS.push({ value: r[0], label: r[1], short: r[2], groupStart: i === 0 });
    });

  var WEEK_BY_VALUE = {};
  WEEKS.forEach(function (o, i) { o.index = i; WEEK_BY_VALUE[o.value] = o; });

  function weekOpt(value) {
    return WEEK_BY_VALUE[String(value)] || WEEK_BY_VALUE.all;
  }

  /* ---------- source links ----------
   * The Sources section of the page is a hand-written list of X posts, each
   * shown as its handle. A note's source is the same thing at bullet scale, so
   * it is parsed the same way: out of the URL. X's oEmbed needs an API key, so
   * there is no fetching the post's text or its author's display name — the
   * handle in the link is all there is, and it is enough to say where a note
   * came from. The server re-parses this and stores nothing it can't verify. */
  var X_POST = /^https?:\/\/(?:www\.)?(?:x|twitter)\.com\/([A-Za-z0-9_]{1,15})\/status(?:es)?\/(\d+)/i;

  function parseXLink(url) {
    var m = X_POST.exec(String(url || '').trim());
    return m ? { url: String(url).trim(), handle: m[1] } : null;
  }

  /* ---------- the store ----------
   * One fetch on load, held in memory. There are a few hundred notes across the
   * league at most, and every team card wants its own slice of them, so the
   * page reads from here rather than going back to the network per team. */

  var notes = [];
  var loaded = false;

  function sortNotes(list) {
    return list.slice().sort(function (a, b) {
      var d = weekOpt(a.week).index - weekOpt(b.week).index;
      if (d) return d;
      // Within a week, the doc's own ordering first (the seeded notes carry it),
      // then the order they were written in.
      var ao = typeof a.order === 'number' ? a.order : 1e9;
      var bo = typeof b.order === 'number' ? b.order : 1e9;
      if (ao !== bo) return ao - bo;
      return String(a.createdAt || '').localeCompare(String(b.createdAt || ''));
    });
  }

  function bunNotesLoad() {
    return fetch(API, { cache: 'no-store' })
      .then(function (r) {
        if (!r.ok) throw new Error('HTTP ' + r.status);
        return r.json();
      })
      .then(function (rows) {
        notes = Array.isArray(rows) ? rows : [];
        loaded = true;
        return true;
      })
      .catch(function () {
        // Left false so the page knows to fall back to the notes committed in
        // data/nfl_projections_2026.json rather than showing a team no notes.
        loaded = false;
        return false;
      });
  }

  function bunNotesLoaded() { return loaded; }

  function bunNotesFor(team, kind) {
    kind = kind || 'note';
    return sortNotes(notes.filter(function (n) {
      return n.team === team && (n.kind || 'note') === kind;
    }));
  }

  // Notes filed with no team: about the league, not about a club.
  function bunNotesLeague() { return bunNotesFor(''); }

  function bunNoteById(id) {
    return notes.filter(function (n) { return n.id === id; })[0] || null;
  }

  function stash(saved) {
    (saved || []).forEach(function (n) {
      var i = -1;
      notes.forEach(function (x, j) { if (x.id === n.id) i = j; });
      if (i === -1) notes.push(n); else notes[i] = n;
    });
  }

  /* ---------- network ---------- */

  function authHeaders(base) {
    var user = typeof getUser === 'function' ? getUser() : null;
    var headers = base || {};
    if (user) headers['X-User-Id'] = user.user_id;
    return headers;
  }

  // Unlike the bets writes, these are not fire-and-forget: a note that silently
  // failed to save is a note the writer believes is published.
  function bunNotesSave(list) {
    return fetch(API, {
      method: 'PUT',
      headers: authHeaders({ 'Content-Type': 'application/json' }),
      body: JSON.stringify({ notes: list })
    }).then(function (r) {
      return r.json().then(function (body) {
        if (!r.ok) throw new Error(body && body.error ? body.error : 'HTTP ' + r.status);
        stash(body.notes);
        return body.notes;
      });
    });
  }

  function bunNotesDelete(id) {
    return fetch(API + '?id=' + encodeURIComponent(id), {
      method: 'DELETE',
      headers: authHeaders()
    }).then(function (r) {
      return r.json().then(function (body) {
        if (!r.ok) throw new Error(body && body.error ? body.error : 'HTTP ' + r.status);
        notes = notes.filter(function (n) { return n.id !== id; });
        return true;
      });
    });
  }

  function newNoteId() {
    return 'n_' + Date.now().toString(36) + Math.random().toString(36).slice(2, 6);
  }

  /* ---------- rendering ----------
   * The bullet is built here rather than on the page so the team card and the
   * league block cannot drift apart, and so the edit affordance is attached in
   * exactly one place. */
  function bunNoteBullet(note) {
    var wk = weekOpt(note.week).short;
    var src = note.source && note.source.url
      ? ' <a class="tc-note-src" href="' + esc(note.source.url) + '" target="_blank" ' +
        'rel="noopener">@' + esc(note.source.handle || 'link') + '</a>'
      : '';
    var edit = admin()
      ? ' <button type="button" class="tc-note-edit" data-note-edit="' + esc(note.id) +
        '" title="Edit note" aria-label="Edit note">' +
        '<svg width="12" height="12" viewBox="0 0 24 24" fill="none" stroke="currentColor" ' +
        'stroke-width="2.5" stroke-linecap="round" stroke-linejoin="round" aria-hidden="true">' +
        '<path d="M12 20h9"></path>' +
        '<path d="M16.5 3.5a2.121 2.121 0 0 1 3 3L7 19l-4 1 1-4 12.5-12.5z"></path></svg></button>'
      : '';
    return '<li data-note-id="' + esc(note.id) + '">' +
      (wk ? '<span class="tc-note-wk">' + esc(wk) + '</span>' : '') +
      esc(note.text) + src + edit + '</li>';
  }

  function bunNoteList(list) {
    if (!list.length) return '';
    return '<ul class="tc-notes">' + list.map(bunNoteBullet).join('') + '</ul>';
  }

  /* ---------- the modal ----------
   * Built once and appended to <body>, like the bet quick-edit modal, so that
   * bullets rendered later — every team card is rendered on open — still find
   * it. A native <dialog>, matching the team card's: Esc, an inert page and a
   * styleable backdrop for free. */

  var dlg, rowsEl, srcInput, srcHint, errEl, titleEl, deleteBtn;
  var teamBtn, teamGrid, teamDd, weekBtn, weekMenu, weekDd;
  var teams = [];
  var draft = { team: '', week: 'all' };
  var editing = null;          // the note being edited, or null when adding
  var onSaved = function () {};

  function ICON_CLOSE() {
    return '<svg width="16" height="16" viewBox="0 0 24 24" fill="none" stroke="currentColor" ' +
      'stroke-width="2" stroke-linecap="round" stroke-linejoin="round" aria-hidden="true">' +
      '<path d="M18 6 6 18"></path><path d="m6 6 12 12"></path></svg>';
  }

  function build() {
    dlg = document.createElement('dialog');
    dlg.className = 'bn-modal';
    dlg.id = 'bn-modal';
    dlg.setAttribute('aria-label', 'Add notes');
    dlg.innerHTML =
      '<div class="bn-head">' +
        '<h4 class="bn-title">Add notes</h4>' +
        '<button type="button" class="bn-close" data-close aria-label="Close">' + ICON_CLOSE() + '</button>' +
      '</div>' +
      '<div class="bn-body">' +
        '<div class="dd-row">' +
          '<div class="dd" data-dd="team">' +
            '<button type="button" class="dd-btn" aria-haspopup="dialog" aria-expanded="false">' +
              '<span class="dd-value">League-wide</span></button>' +
            '<div class="dd-pop dd-overlay" role="dialog" aria-modal="true" aria-label="Team" hidden>' +
              '<div class="tp"></div></div>' +
          '</div>' +
          // Both pickers open over the page rather than under their trigger.
          // The team grid always did; the week list has to here, because an
          // anchored panel inside this dialog is clipped by the body's own
          // scroller — a fixed overlay is clipped by nothing.
          '<div class="dd" data-dd="week">' +
            '<button type="button" class="dd-btn" aria-haspopup="dialog" aria-expanded="false">' +
              '<span class="dd-value">General — no week</span></button>' +
            '<div class="dd-pop dd-overlay" role="dialog" aria-modal="true" aria-label="Week" hidden>' +
              '<div class="dd-menu" role="menu"></div></div>' +
          '</div>' +
        '</div>' +
        '<div class="bn-rows"></div>' +
        '<label class="bn-label" for="bn-src">Source <span>optional &mdash; an X post</span></label>' +
        '<input type="url" id="bn-src" class="bn-input" placeholder="https://x.com/handle/status/…" ' +
          'autocomplete="off" spellcheck="false">' +
        '<p class="bn-hint"></p>' +
      '</div>' +
      '<p class="bn-error" hidden></p>' +
      '<div class="bn-actions">' +
        '<button type="button" class="bn-btn is-danger" data-act="delete" hidden>Delete</button>' +
        '<button type="button" class="bn-btn" data-close>Cancel</button>' +
        '<button type="button" class="bn-btn is-primary" data-act="save">Save</button>' +
      '</div>';
    document.body.appendChild(dlg);

    titleEl = dlg.querySelector('.bn-title');
    rowsEl = dlg.querySelector('.bn-rows');
    srcInput = dlg.querySelector('#bn-src');
    srcHint = dlg.querySelector('.bn-hint');
    errEl = dlg.querySelector('.bn-error');
    deleteBtn = dlg.querySelector('[data-act="delete"]');

    var teamRoot = dlg.querySelector('[data-dd="team"]');
    var weekRoot = dlg.querySelector('[data-dd="week"]');
    teamBtn = teamRoot.querySelector('.dd-btn');
    teamGrid = teamRoot.querySelector('.tp');
    weekBtn = weekRoot.querySelector('.dd-btn');
    weekMenu = weekRoot.querySelector('.dd-menu');

    buildWeekMenu(weekMenu, WEEKS);
    teamDd = registerDropdown(teamRoot);
    weekDd = registerDropdown(weekRoot);

    weekMenu.addEventListener('click', function (e) {
      var opt = e.target.closest ? e.target.closest('.dd-opt') : null;
      if (!opt) return;
      draft.week = opt.getAttribute('data-week');
      closeDropdown(weekDd, true);
      syncDraft();
    });

    teamGrid.addEventListener('click', function (e) {
      var btn = e.target.closest ? e.target.closest('.tp-team, .tp-any') : null;
      if (!btn) return;
      draft.team = btn.getAttribute('data-team') || '';
      closeDropdown(teamDd, true);
      syncDraft();
    });

    dlg.addEventListener('click', function (e) {
      if (e.target === dlg || e.target.closest('[data-close]')) { close(); return; }
      var act = e.target.closest('[data-act]');
      if (act && act.dataset.act === 'save') save();
      if (act && act.dataset.act === 'delete') removeNote();
      var add = e.target.closest('[data-row-add]');
      if (add) addRow('', true);
      var del = e.target.closest('[data-row-del]');
      if (del) { del.closest('.bn-row').remove(); syncRows(); }
    });

    // Escape belongs to whichever picker is open before it belongs to the
    // dialog — otherwise closing the team grid closes the whole note with it.
    dlg.addEventListener('cancel', function (e) {
      if (anyDropdownOpen()) { e.preventDefault(); closeAllDropdowns(); }
    });

    dlg.addEventListener('close', function () {
      ScrollLock.unlock();
      closeAllDropdowns();
    });

    rowsEl.addEventListener('keydown', function (e) {
      if (e.key !== 'Enter') return;
      e.preventDefault();
      if (e.metaKey || e.ctrlKey || editing) { save(); return; }
      addRow('', true);
    });

    srcInput.addEventListener('input', syncSource);
  }

  /* One note per row. The first row carries the "+" that adds another, because
   * that is where a writer already is when they realise they have a second
   * thing to say; every row after it carries the "×" that takes it back out. */
  function rowHtml(text, first) {
    return '<div class="bn-row">' +
      '<input type="text" class="bn-input" value="' + esc(text) + '" ' +
        'placeholder="' + (first ? 'Write a note…' : 'And another…') + '" ' +
        'autocomplete="off">' +
      (first
        ? '<button type="button" class="bn-row-btn" data-row-add title="Add another note" ' +
          'aria-label="Add another note">+</button>'
        : '<button type="button" class="bn-row-btn" data-row-del title="Remove this note" ' +
          'aria-label="Remove this note">&times;</button>') +
      '</div>';
  }

  function addRow(text, focus) {
    rowsEl.insertAdjacentHTML('beforeend', rowHtml(text || '', !rowsEl.children.length));
    if (focus) rowsEl.lastElementChild.querySelector('input').focus();
  }

  // After a removal the surviving rows are renumbered, so that the "+" is
  // always on the first one.
  function syncRows() {
    var rows = rowsEl.querySelectorAll('.bn-row');
    if (!rows.length) { addRow('', false); return; }
    var texts = [].map.call(rows, function (r) { return r.querySelector('input').value; });
    rowsEl.innerHTML = texts.map(function (t, i) { return rowHtml(t, i === 0); }).join('');
  }

  function syncDraft() {
    syncTeamPicker(teamBtn, teamGrid, draft.team, teams, { emptyLabel: 'League-wide' });
    syncWeekMenu(weekBtn, weekMenu, draft.week, weekOpt(draft.week).label);
  }

  function syncSource() {
    var raw = srcInput.value.trim();
    if (!raw) { srcHint.textContent = ''; srcHint.className = 'bn-hint'; return; }
    var parsed = parseXLink(raw);
    srcHint.textContent = parsed
      ? '@' + parsed.handle
      : 'Not an X post link — expected x.com/handle/status/…';
    srcHint.className = 'bn-hint' + (parsed ? ' is-ok' : ' is-bad');
  }

  function fail(message) {
    errEl.textContent = message;
    errEl.hidden = !message;
  }

  function open(opts) {
    opts = opts || {};
    editing = opts.note || null;
    fail('');

    draft.team = editing ? (editing.team || '') : (opts.team || '');
    draft.week = editing ? (editing.week || 'all') : (opts.week || 'all');

    titleEl.textContent = editing ? 'Edit note' : 'Add notes';
    dlg.setAttribute('aria-label', titleEl.textContent);
    deleteBtn.hidden = !editing;

    rowsEl.innerHTML = '';
    addRow(editing ? editing.text : '', false);

    srcInput.value = editing && editing.source ? editing.source.url : '';
    syncSource();
    syncDraft();

    dlg.showModal();
    // Taken after showModal so a double-open — which throws — cannot take a
    // lock nothing will release. See scripts/base/scroll-lock.js.
    ScrollLock.lock();
    rowsEl.querySelector('input').focus();
  }

  function close() {
    if (dlg.open) dlg.close();
  }

  function save() {
    var texts = [].map.call(rowsEl.querySelectorAll('input'), function (i) {
      return i.value.trim();
    }).filter(Boolean);

    if (!texts.length) { fail('Write a note first.'); return; }

    var raw = srcInput.value.trim();
    var source = raw ? parseXLink(raw) : null;
    if (raw && !source) { fail('That source is not an X post link.'); return; }

    // One source for the batch: several bullets saved together came off the
    // same post. Editing one note edits that note's own link.
    var payload = texts.map(function (text) {
      var note = {
        id: editing ? editing.id : newNoteId(),
        team: draft.team,
        week: draft.week,
        kind: editing ? (editing.kind || 'note') : 'note',
        text: text
      };
      if (editing && editing.createdAt) note.createdAt = editing.createdAt;
      if (editing && typeof editing.order === 'number') note.order = editing.order;
      if (source) note.source = source;
      return note;
    });

    fail('');
    dlg.classList.add('is-saving');
    bunNotesSave(payload)
      .then(function () {
        dlg.classList.remove('is-saving');
        close();
        onSaved();
      })
      .catch(function (e) {
        dlg.classList.remove('is-saving');
        fail(e.message || 'Could not save.');
      });
  }

  function removeNote() {
    if (!editing) return;
    if (!confirm('Delete this note? This cannot be undone.')) return;
    dlg.classList.add('is-saving');
    bunNotesDelete(editing.id)
      .then(function () {
        dlg.classList.remove('is-saving');
        close();
        onSaved();
      })
      .catch(function (e) {
        dlg.classList.remove('is-saving');
        fail(e.message || 'Could not delete.');
      });
  }

  /* ---------- wiring ----------
   * Called by the page once its team list has landed. Everything admin-only —
   * the floating button, the per-bullet pencils — is wired here and simply
   * never built for anyone else; the server is what actually refuses the write.
   */
  function initBunNotes(opts) {
    opts = opts || {};
    teams = opts.teams || [];
    onSaved = opts.onSaved || function () {};
    // Nothing that writes is offered while the store is unreachable: the page is
    // showing the JSON fallback then, and a note saved against it would vanish
    // into a list the reader cannot see.
    if (!admin() || !loaded) return;

    build();
    buildTeamPicker(teamGrid, teams, { anyLabel: 'League-wide — no team' });

    var fab = opts.fab || document.getElementById('bn-fab');
    if (fab) {
      fab.hidden = false;
      fab.addEventListener('click', function () { open({}); });
    }

    // Delegated off document: the pencils live inside a team card that is
    // rebuilt every time one opens, and the "add a note" buttons with them.
    document.addEventListener('click', function (e) {
      var edit = e.target.closest ? e.target.closest('[data-note-edit]') : null;
      if (edit) {
        var note = bunNoteById(edit.getAttribute('data-note-edit'));
        if (note) { if (opts.beforeOpen) opts.beforeOpen(); open({ note: note }); }
        return;
      }
      var add = e.target.closest ? e.target.closest('[data-note-add]') : null;
      if (add) {
        if (opts.beforeOpen) opts.beforeOpen();
        open({ team: add.getAttribute('data-note-add') || '' });
      }
    });
  }

  global.BUN_WEEKS = WEEKS;
  global.parseXLink = parseXLink;
  global.bunNotesLoad = bunNotesLoad;
  global.bunNotesLoaded = bunNotesLoaded;
  global.bunNotesFor = bunNotesFor;
  global.bunNotesLeague = bunNotesLeague;
  global.bunNoteById = bunNoteById;
  global.bunNoteBullet = bunNoteBullet;
  global.bunNoteList = bunNoteList;
  global.bunNotesSave = bunNotesSave;
  global.bunNotesDelete = bunNotesDelete;
  global.initBunNotes = initBunNotes;
  global.openBunNoteModal = open;
})(window);
